Improve the UX of Open Wallet dialog

wukong1971 opened the pull request:
This PR addresses the first two items in the "Next step improvements" of #932:

1. Show all the error messages, like "Wrong Password", "File not exist", "Invalid wallet file" etc., within the same dialog, and allow the user to make changes to their decisions in the same dialog when these errors occur.

![image](https://user-images.githubusercontent.com/87334822/132820190-1cbbf5fd-4d14-4317-a287-3aec378c4a41.png)

![image](https://user-images.githubusercontent.com/87334822/132820240-0a98e50a-43fe-4449-89e6-093af1f3533c.png)

2. Update the UX of menu action "Wallet -> Load" to also use this new Open Wallet dialog.
In this update, I also renamed the current menu item "Load" to "Open", and added a shortcut "Ctrl+O" for this menu item, which I believe is more consistent with the expectation of the UI from most users.

wukong1971 commented on scripts/joinmarket-qt.py:
```diff
@@ -2003,19 +2021,22 @@ def selectWallet(self, testnet_seed=None):
             self.loadWalletFromBlockchain(firstarg, pwd)
 
-    def loadWalletFromBlockchain(self, firstarg=None, pwd=None):
+    def loadWalletFromBlockchain(self, firstarg=None, pwd=None, rethrow=False):
```
This `rethrow` flag is a temporarily solution to keep the behavior of "Generate Wallet" the same as before, while allowing the error message to be displayed directly in the "Open Wallet Dialog".

Eventually, when we have fully completed our redesign of the open/create wallet workflow, this flag would no longer be necessary and will be removed at that point.

kristapsk commented:
> added a shortcut "Ctrl+O"

Proper approach is to use `QKeySequence.Open` instead of hardcoded "Ctrl+O". https://doc.qt.io/qtforpython-5/PySide2/QtGui/QKeySequence.html#standard-shortcuts

wukong1971 commented:
> Proper approach is to use `QKeySequence.Open` instead of hardcoded "Ctrl+O". https://doc.qt.io/qtforpython-5/PySide2/QtGui/QKeySequence.html#standard-shortcuts

I did "Ctrl + O" because there was a "Ctrl + Q" in place already.

Just pushed a new commit to replace both with `QKeySequence`

kristapsk commented:
> there was a "Ctrl + Q" in place already

I didn't know about `QKeySequence` back then.

kristapsk commented on scripts/joinmarket-qt.py:
```diff
@@ -1632,9 +1632,10 @@ def closeEvent(self, event):
-        loadAction.setStatusTip('Load wallet from file')
-        loadAction.triggered.connect(self.selectWallet)
+        openWalletAction = QAction('&Open...', self)
+        openWalletAction.setStatusTip('Open joinmarket wallet file')
```
nit - it's proper noun, so should be either "Joinmarket" or "JoinMarket" (my preference).
```suggestion
        openWalletAction.setStatusTip('Open JoinMarket wallet file')
```

kristapsk commented:
Another possible UX improvement - by default dialog is focused on "Passphrase" input box. But after I press "Choose..." and select wallet file, it loses that focus. Which means unnecessary extra click with mouse or pressing Tab multiple times (Tab order of fields also seems to be random).

wukong1971 commented:
Addressed all PR comments, PTAL, thanks!

PulpCattel commented:
I like the concept and it seems to work nicely.
One thing I noticed, the first time the `open wallet dialog` opens it shows the text cursor, after an error is thrown, the new dialog has no cursor. Clicking choose, or re-loading the dialog, restores the cursor.

wukong1971 commented:
> One thing I noticed, the first time the `open wallet dialog` opens it shows the text cursor, after an error is thrown, the new dialog has no cursor. Clicking choose, or re-loading the dialog, restores the cursor.

Right, I've noticed this issue as well, but I couldn't figure out how to fix it.
If anyone has any suggestions / advice, I'd be happy to try them out.

AdamISZ commented:
My guess is that calling `.exec_()` twice on the same dialog window is the reason for the error (but I emphasize, *guess*).

I find it can be solved with this edit:

```python3
    def openWallet(self):
        walletLoaded = False
        errortext = ""
        while not walletLoaded:
            openWalletDialog = JMOpenWalletDialog()
            openWalletDialog.errorMessageLabel.setText(errortext)
            if openWalletDialog.exec_() == QDialog.Accepted:
                wallet_path = openWalletDialog.walletFileEdit.text()
                if not os.path.isabs(wallet_path):
                    wallet_path = os.path.join(jm_single().datadir, 'wallets', wallet_path)
                
                try:
                    walletLoaded = mainWindow.loadWalletFromBlockchain(wallet_path, openWalletDialog.passphraseEdit.text(), rethrow=True)
                except Exception as e:
                    errortext = str(e)
            else:
                break
```
in other words, edit the text as a variable that persists over each iteration of the while loop and set the errorMessageLabel to that.

Since that seems to work, I suppose my guess is vaguely correct but I doubt it matters.

I'd be happy to merge this once that's fixed, I think. Taking another look at the code now just in case.

AdamISZ commented:
It is slightly counter-intuitive that if you get a 'wrong password' error, the filename reverts to the default `wallet.jmdat`, no matter if you actually entered a different wallet name. But it's not a big deal.

AdamISZ commented:
tACK modulo adding the patch as per https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1016#issuecomment-961363773 or something similar.

wukong1971 commented:
> It is slightly counter-intuitive that if you get a 'wrong password' error, the filename reverts to the default `wallet.jmdat`, no matter if you actually entered a different wallet name. But it's not a big deal.

I can't reproduce this error on my branch, I think this is a new issue introduced by https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1016#issuecomment-961363773

wukong1971 commented:
Updated this PR based on the suggestion of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1016#issuecomment-961363773.

Also fixed the problem mentioned in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1016#issuecomment-961367456 by adding a new `wallet_file_text` variable.

Merged latest master.

kristapsk commented:
tACK, looks good to me. Commits need to be squashed before merge into master.

wukong1971 commented:
Commits squashed.

kristapsk reviewed (approved):
ACK fba9cc3d8f7b69c2609bf551c459771eddb97369

kristapsk commented on jmqtui/jmqtui/open_wallet_dialog.py:
```diff
@@ -92,6 +92,30 @@ def setupUi(self, OpenWalletDialog):
+
+        self.horizontalLayout_4.addWidget(self.errorMessageLabel)
+
+
```
nit - unnecessary extra blank line

wukong1971 replied:
This file is auto generated, we're not supposed to modify this file.

kristapsk commented on scripts/joinmarket-qt.py:
```diff
@@ -1960,6 +1963,28 @@ def recoverWallet(self):
+                wallet_path = wallet_file_text
+                if not os.path.isabs(wallet_path):
+                    wallet_path = os.path.join(jm_single().datadir, 'wallets', wallet_path)
+                
```
nit - remove extra whitespace
```suggestion

```

wukong1971 replied:
Done.

kristapsk reviewed (commented):
Two small nits, otherwise I'm ready to merge this.

wukong1971 commented:
Comments addressed, PTAL :)

kristapsk reviewed (approved):
re-ACK c706c6e1e9cbe7a89b04a7d01ab0907cd0ab880a
